### PR TITLE
Add support for classes with hidden fields #563

### LINKED ISSUE #563: Errors in test generation when there is variable hiding in src code
**Description**

When we generate tests for a class with a field that hides field from superclass, incorrect tests are generated.

**To Reproduce**

Launch the plugin on `class C` in the following example:

```Java
class A {
    public int f;
    void setF(int val) { this.f = val; }
    int getF() { return this.f; }
}

class B extends A {
    public int f;
}

public class C {
    boolean foo(B b) {
        return b.getF() == b.f + 1;
    }
    boolean bar(B b) {     // replace body with return b.getF() == b.f and test generation will work ok
        if (b.getF() == b.f)  
            return true;
        else
            return false;
    }
}

```

**Expected behavior**

Correct tests with complete coverage are generated.

**Actual behavior**

For function `bar` plugin generates incorrect test `testBar_BGetFEqualsBF` (assertEquals in this test fails when launching).
For function `foo` there is no testcase in which return value would be true.

Both problems disappear if we rename field in `class A` to any other name.

**Visual proofs (screenshots, logs, images)**

Incorrect test generated:
```Java
    @Test
    @DisplayName("bar: BGetF -> return true")
    public void testBar_BGetFEqualsBF() {
        C c = new C();
        B b = new B();
        b.f = 1;

        boolean actual = c.bar(b);

        assertFalse(actual);

        int finalBF = b.f;

        assertEquals(0, finalBF);
    }
```

**Environment**

plugin, Java11

**Additional context**

The problem also happens if we mark the field in `class B` as private: in some of the generated tests there will be accesses to this private member, so generated tests are not even compilable. Try launching test generation on `C.baz()` in this example:
```Java
package inheritanceIssues;

class A {
    public int f;
    void setF(int val) { this.f = val; }
    int getF() { return this.f; }
}

class B extends A {
    private int f;
}

public class C {
    boolean baz(B b) {
        return b.getF() == 1;
    }
}
```

### PR DESCRIPTION
**Pull request by volivan239**
# Description

Now if class has hidden fields, they can be recognized by engine and correct code for instantiation of such classes will be generated.

Fixes #563

## Type of Change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

## Automated Testing

Switched `org.utbot.examples.objects.HiddenFieldExampleTest#testCheckSuccField` to enabled, added `org.utbot.examples.objects.HiddenFieldAccessModifiersTest`

## Manual Scenario 

Tested plugin on examples from #563 -- generates tests the way as expected.

# Checklist (remove irrelevant options):

_This is the author self-check list_

- [ ] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [ ] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [ ] No new warnings
- [x] New tests have been added
- [ ] All tests pass locally with my changes

### CODE REVIEW

File: utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt [hunk 1]
```diff
@@ -7,22 +7,9 @@ import java.lang.reflect.Method
...
-        } catch (e: NoSuchFieldException) {
-            try {
-                it.getDeclaredField(name)
-            } catch (e: NoSuchFieldException) {
-                null
-            }
-        }
-    }
-    .firstOrNull()
+fun Class<*>.findDirectAccessedField(fieldName: String): Field? = generateSequence(this) { it.superclass }
```
Comment (CaelmBleidd) [inline utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt:10]: Please, add documentation here. Especially about what 'directAccessed' means

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt [hunk 2]
```diff
@@ -855,7 +856,7 @@ open class FieldId(val declaringClass: ClassId, val name: String) {
         return result
     }
 
-    override fun toString() = declaringClass.findFieldOrNull(name).toString()
+    override fun toString() = declaringClass.fieldOrNull(this).toString()
```
Comment (CaelmBleidd) [inline utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt:859]: Why not `findField` anymore?

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt [hunk 3]
```diff
@@ -286,9 +285,17 @@ val ClassId.isMap: Boolean
 val ClassId.isIterableOrMap: Boolean
     get() = isIterable || isMap
 
-fun ClassId.findFieldOrNull(fieldName: String): Field? = jClass.findFieldOrNull(fieldName)
+fun ClassId.fieldOrNull(fieldId: FieldId): Field? {
+    if (this.isNotSubtypeOf(fieldId.declaringClass))
+        return null
```
Comment (CaelmBleidd) [inline utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt:290]: Please, do not omit brackets for multiline constructions. Minor: redundant `this`

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt [hunk 4]
```diff
@@ -286,9 +285,17 @@ val ClassId.isMap: Boolean
...
+    return try {
+        fieldId.field
+    } catch (e: IllegalStateException) {
+        null
```
Comment (CaelmBleidd) [inline utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt:294]: I believe, it is exception-dependent logic. Why you can't do it without throwing an exception? It is not a cheap operation after all. Moreover, please, add an empty line before the return

File: utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt (hunk 1 above)
Comment (CaelmBleidd) [inline utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt:10]: Why it is not `orNull` anymore?

File: utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt [hunk 5]
```diff
@@ -28,10 +28,11 @@ class Hierarchy(private val typeRegistry: TypeRegistry) {
         type as? RefType ?: error("$type is not a refType")
 
         val realType = typeRegistry.findRealType(type) as RefType
+        val realFieldDeclaringClass = typeRegistry.findRealType(field.declaringClass.type) as RefType
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt:31]: It is a declaringType

File: utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt [hunk 6]
```diff
@@ -28,10 +28,11 @@ class Hierarchy(private val typeRegistry: TypeRegistry) {
...
 
-        val ancestorType = ancestors(realType.sootClass.id).lastOrNull { it.declaresField(field.subSignature) }?.type
-            ?: error("No such field ${field.subSignature} found in ${realType.sootClass.name}")
-        return ChunkId("$ancestorType", field.name)
+        if (realFieldDeclaringClass.sootClass !in ancestors(realType.sootClass.id))
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt:33]: Please, do not omit brackets

Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt:33]: Why do we always have such a field in ancestors?

File: utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt (hunk 5 above)
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt:31]: Why did it work earlier? Please, add corresponding comment here with explanation and example

File: utbot-framework/src/main/kotlin/org/utbot/engine/Strings.kt [hunk 7]
```diff
@@ -313,7 +313,7 @@ sealed class UtAbstractStringBuilderWrapper(className: String) : BaseOverriddenW
 
         val arrayValuesChunkId = typeRegistry.arrayChunkId(charArrayType)
 
-        val valuesFieldChunkId = hierarchy.chunkIdForField(utStringClass.type, overriddenClass.valueField)
+        val valuesFieldChunkId = hierarchy.chunkIdForField(utStringClass.type, utStringClass.valueField)
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Strings.kt:316]: Are you sure? Why? 

Moreover, if it is true, do we write in this field in all other places in the wrapper? Not only in the resolver

File: utbot-framework/src/main/kotlin/org/utbot/engine/Traverser.kt [hunk 8]
```diff
@@ -639,7 +639,7 @@ class Traverser(
             SECURITY_FIELD_SIGNATURE -> SecurityManager()
             FIELD_FILTER_MAP_FIELD_SIGNATURE -> mapOf(Reflection::class to arrayOf("fieldFilterMap", "methodFilterMap"))
             METHOD_FILTER_MAP_FIELD_SIGNATURE -> emptyMap<Class<*>, Array<String>>()
-            else -> declaringClass.id.jClass.findField(field.name).let { it.withAccessibility { it.get(null) } }
+            else -> field.fieldId.field.let { it.withAccessibility { it.get(null) } }
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/Traverser.kt:642]: Obviously, a problem with naming: `field.fieldId.field`

Moreover, now we have unused `declaringClass` variable

File: utbot-framework/src/main/kotlin/org/utbot/engine/ValueConstructor.kt [hunk 9]
```diff
@@ -214,7 +214,7 @@ class ValueConstructor {
 
         model.fields.forEach { (field, fieldModel) ->
             val declaredField =
-                javaClass.findFieldOrNull(field.name) ?: error("Can't find field: $field for $javaClass")
+                model.classId.fieldOrNull(field) ?: error("Can't find field: $field for $javaClass")
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/engine/ValueConstructor.kt:217]: Returning back to the `fieldOrNull` implementation, here it looks like: 
`if we don't have a field, throw an exception. I'll catch it, return null and throw an exception once again`

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 10]
```diff
@@ -128,21 +127,17 @@ internal class CgVariableConstructor(val context: CgContext) :
...
             // Example:
             // Object heapByteBuffer = createInstance("java.nio.HeapByteBuffer");
             // branchRegisterRequest.byteBuffer = heapByteBuffer;
             // byteBuffer is field of type ByteBuffer and upper line is incorrect
-            val canFieldBeDirectlySetByVariableAndFieldTypeRestrictions =
-                fieldFromVariableSpecifiedType != null && fieldFromVariableSpecifiedType.type.id == variableForField.type
-            if (canFieldBeDirectlySetByVariableAndFieldTypeRestrictions && fieldId.canBeSetIn(testClassPackageName)) {
-                // TODO: check if it is correct to use declaringClass of a field here
-                val fieldAccess = if (field.isStatic) CgStaticFieldAccess(fieldId) else CgFieldAccess(obj, fieldId)
+            if (field != null && field.type.id == variableForField.type && fieldId.canBeSetIn(testClassPackageName)) {
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:139]: @Damtev or @EgorkaKulikov, please, look over these changes with code generation

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt [hunk 11]
```diff
@@ -72,7 +76,10 @@ fun stringLiteral(string: String) = CgLiteral(stringClassId, string)
 
 // non-static fields
 operator fun CgExpression.get(fieldId: FieldId): CgFieldAccess =
-    CgFieldAccess(this, fieldId)
+    if (type.jClass.findDirectAccessedField(fieldId.name) != fieldId.field)
+        CgFieldAccess(CgTypeCast(fieldId.declaringClass, this), fieldId)
+    else
+        CgFieldAccess(this, fieldId)
```
Comment (CaelmBleidd) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt:82]: brackets

File: utbot-framework/src/test/kotlin/org/utbot/examples/UtModelTestCaseChecker.kt [hunk 12]
```diff
@@ -167,8 +165,8 @@ internal abstract class UtModelTestCaseChecker(
     /**
      * Finds field model in [UtCompositeModel] and [UtAssembleModel]. For assemble model supports direct field access only.
      */
-    protected fun UtModel.findField(fieldName: String): UtModel =
-        findField(this.classId.jClass.findField(fieldName).fieldId)
+    protected fun UtModel.findField(fieldName: String, declarationClass: ClassId = this.classId): UtModel =
```
Comment (CaelmBleidd) [inline utbot-framework/src/test/kotlin/org/utbot/examples/UtModelTestCaseChecker.kt:168]: is `declaration` the right word?

File: utbot-sample/src/main/java/org/utbot/examples/objects/HiddenFieldSuccClass.java [hunk 13]
```diff
@@ -2,4 +2,9 @@
 
 public class HiddenFieldSuccClass extends HiddenFieldSuperClass {
     public double b;
+
+    @Override
+    public String toString() {
+        return b + ". Super b: " + super.b;
+    }
```
Comment (CaelmBleidd) [inline utbot-sample/src/main/java/org/utbot/examples/objects/HiddenFieldSuccClass.java:9]: Isn't it a forgotten method implementation?

File: utbot-framework/src/test/kotlin/org/utbot/examples/objects/HiddenFieldAccessModifiersTest.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+internal class HiddenFieldAccessModifiersTest : UtValueTestCaseChecker(testClass = HiddenFieldAccessModifiersExample::class) {
+    @Test
+    fun testCheckSuperFieldEqualsOne() {
+        check(
+            HiddenFieldAccessModifiersExample::checkSuperFieldEqualsOne,
+            eq(3),
+            { o, _ -> o == null },
+            { _, r -> r == true },
+            { _, r -> r == false},
+            coverage = DoNotCalculate
```
Comment (CaelmBleidd) [inline utbot-framework/src/test/kotlin/org/utbot/examples/objects/HiddenFieldAccessModifiersTest.kt:17]: Why?

**Review by CaelmBleidd — CHANGES_REQUESTED** [review]

File: utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt (hunk 1 above)
Comment (volivan239) [reply utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt]: Added documentation for this method

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt (hunk 2 above)
Comment (volivan239) [reply utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt]: Previously it was a method that was finding `Field` by its name. Now, when the `FieldId` is passed as argument, it is not really finding -- it's more like checking that this field belongs to this class and returning `Field`.

Renamed to `fieldByIdOrNull` for greater clarity

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt (hunk 3 above)
Comment (volivan239) [reply utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt]: fixed

File: utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt (hunk 4 above)
Comment (volivan239) [reply utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/util/IdUtil.kt:293]: fixed, although it's a little bit copy-pasted now

File: utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt (hunk 1 above)
Comment (volivan239) [reply utbot-core/src/main/kotlin/org/utbot/common/KClassUtil.kt]: Renamed to `findDirectAccessedFieldOrNull`

File: utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt (hunk 5 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt]: Renamed to `realFieldDeclraingType`. Is "declaring type" a correct term?

File: utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt (hunk 6 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt]: fixed here

Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/engine/Hierarchy.kt]: Are there other ways how field can belong to the class? At least, similar check was for `ancestorType` before these changes

File: utbot-framework/src/main/kotlin/org/utbot/engine/Traverser.kt (hunk 8 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/engine/Traverser.kt]: Just rewrote using `declaringClass`, seems better now

File: utbot-framework/src/main/kotlin/org/utbot/engine/ValueConstructor.kt (hunk 9 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/engine/ValueConstructor.kt]: Added same method without `orNull`, fixed

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt (hunk 11 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt]: fixed here

File: utbot-framework/src/test/kotlin/org/utbot/examples/UtModelTestCaseChecker.kt (hunk 12 above)
Comment (volivan239) [reply utbot-framework/src/test/kotlin/org/utbot/examples/UtModelTestCaseChecker.kt]: replaced with `declaring`

File: utbot-framework/src/test/kotlin/org/utbot/examples/objects/HiddenFieldAccessModifiersTest.kt (hunk 14 above)
Comment (volivan239) [reply utbot-framework/src/test/kotlin/org/utbot/examples/objects/HiddenFieldAccessModifiersTest.kt]: added coverage here and to `org.utbot.examples.objects.HiddenFieldExampleTest#testCheckSuccField`

File: utbot-sample/src/main/java/org/utbot/examples/objects/HiddenFieldSuccClass.java (hunk 13 above)
Comment (volivan239) [reply utbot-sample/src/main/java/org/utbot/examples/objects/HiddenFieldSuccClass.java]: This override really seems useless, removed it

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt (hunk 10 above)
Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:137]: There are no functional changes, but I am a bit confused with this `TODO` removing. Are you sure it is resolved?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt [hunk 15]
```diff
@@ -72,7 +76,11 @@ fun stringLiteral(string: String) = CgLiteral(stringClassId, string)
 
 // non-static fields
 operator fun CgExpression.get(fieldId: FieldId): CgFieldAccess =
-    CgFieldAccess(this, fieldId)
+    if (type.jClass.findDirectAccessedFieldOrNull(fieldId.name) != fieldId.field) {
+        CgFieldAccess(CgTypeCast(fieldId.declaringClass, this), fieldId)
+    } else {
+        CgFieldAccess(this, fieldId)
+    }
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt:83]: This functionality was just a shortcut for a `CgFieldAccess` constructor. I don't think we should use more complex logic with additional checks here.

Also, this logic does not always work - there is no check that `fieldId.declaringClass` is accessible, so there is possible situation with cast to a private class.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/CgKotlinRenderer.kt [hunk 16]
```diff
@@ -121,7 +121,9 @@ internal class CgKotlinRenderer(context: CgContext, printer: CgPrinter = CgPrint
     // Property access
 
     override fun visit(element: CgFieldAccess) {
+        if (element.caller is CgTypeCast) print("(")
         element.caller.accept(this)
+        if (element.caller is CgTypeCast) print(")")
```
Comment (Damtev) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/CgKotlinRenderer.kt:126]: Why?

**Review by Damtev — CHANGES_REQUESTED** [review]

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/CgKotlinRenderer.kt (hunk 16 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/CgKotlinRenderer.kt:126]: If lhs is typecast, we need brackets for correct priority. Without this, the generated code will be like `Succ as Super.f` and not `(Succ as Super).f`.

This is a little bit of a hack, but I didn't find more general solution.

Comment (Damtev) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/CgKotlinRenderer.kt:126]: This is definitely a hack, which is also not total - it does not cover similar cases for a method invocation and an array access. We need to find another solution.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt (hunk 15 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/DslUtil.kt:83]: What is the best place to move this logic to? Before calling this there are checks that `fieldId` is accessible, aren't they enough?

Also, I believe that this logic should solve the removed TODO: it checks if we should use obj or declaring class as a caller. But maybe I misunderstood the TODO

**Comment by volivan239** [comment]
After discussion with @Damtev and @ArsenHD it was decided to merge into main only changes concerning engine (#627) so far. It was also decided to leave codegen untouched until @ArsenHD finishes his changes in codegen so we would have more rational ways of dealing with field hiding there.

**Comment by volivan239** [comment]
Corresponding issue was finally solved in #763, closing this one therefore.

**Pull request closed without merging**